include/nbl/builtin/hlsl/math/complex.hlsl

```diff
@@ -113,32 +108,54 @@ struct complex_t
     // ----------------- Relational operators -----------------------------
     bool operator==(const complex_t rhs)
     {
-            return !(uint64_t(m_real) ^ uint64_t(rhs.m_real)) && !(uint64_t(m_imag) ^ uint64_t(rhs.m_imag));
+            return m_real == rhs.m_real && m_imag == rhs.m_imag;
     }
     bool operator!=(const complex_t rhs)
     {
-            return uint64_t(m_real) ^ uint64_t(rhs.m_real) || uint64_t(m_imag) ^ uint64_t(rhs.m_imag);
+            return m_real != rhs.m_real || m_imag != rhs.m_imag;
     }
-            
 };
 
-// ---------------------- Compound assignment functors ----------------------
-template <typename Scalar>
-struct ComplexCompoundHelper{              
-    using add = assign_add_t<complex_t<Scalar> >;
-    using subtract = assign_subtract_t<complex_t<Scalar> >;
-    using mul = assign_mul_t<complex_t<Scalar> >;
-    using div = assign_div_t<complex_t<Scalar> >;
+// ---------------------- Compound assign operators ------------------------
+// Specializations of the structs found in functional.hlsl
+// TODO: Figure out a way of making them have an identity (const static member struct doesn't compile on Godbolt)
+
+template<typename Scalar> 
+struct plus_assign< complex_t<Scalar> > {
+    void operator()(NBL_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(complex_t<Scalar>) rhs) {
+        lhs = lhs + rhs;                                                             
+    }                                          
+};
+
+template<typename Scalar> 
+struct multiplies_assign< complex_t<Scalar> > {
+    void operator()(NBL_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(complex_t<Scalar>) rhs) {
+        lhs = lhs * rhs;                                                             
+    }
+
+    void operator()(NBL_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(Scalar) rhs) {
+        lhs = lhs * rhs;                                                             
+    }                                          
+};
```

Comment:
Ok I see you need to specialize because HLSL has no conversion operators that actually work 🤦 

However, you should specialize the non `_assign` and make the `_assign` use not the +-/* operators, but the non-assigns themselves

Reply:
I specialize both for complex to support multiplication/division by scalars with less FMAs

Reply:
Alternatively I can try making the non-assign have a templated operator() on a different type

Reply:
if you implement `X_assign` in terms of `X` you only need to specialize the `X`